strings/base_string.h

```diff
@@ -1,34 +1,143 @@
 
 namespace winrt::impl
 {
-    inline void* duplicate_string(void* other)
+    struct atomic_ref_count
     {
-        void* result = nullptr;
-        check_hresult(WINRT_WindowsDuplicateString(other, &result));
-        return result;
+        atomic_ref_count() noexcept = default;
+
+        explicit atomic_ref_count(uint32_t count) noexcept : m_count(count)
+        {
+        }
+
+        uint32_t operator=(uint32_t count) noexcept
+        {
+            return m_count = count;
+        }
+
+        uint32_t operator++() noexcept
+        {
+            return m_count.fetch_add(1, std::memory_order_relaxed) + 1;
+        }
+
+        uint32_t operator--() noexcept
+        {
+            auto const remaining = m_count.fetch_sub(1, std::memory_order_release) - 1;
+
+            if (remaining == 0)
+            {
+                std::atomic_thread_fence(std::memory_order_acquire);
+            }
+
+            return remaining;
+        }
+
+        operator uint32_t() const noexcept
+        {
+            return m_count;
+        }
+
+    private:
+
+        std::atomic<uint32_t> m_count;
+    };
+
+    constexpr uint32_t hstring_reference_flag{ 1 };
+
+    struct hstring_header
+    {
+        uint32_t flags;
+        uint32_t length;
+        uint32_t padding1;
+        uint32_t padding2;
+        wchar_t const* ptr;
+    };
+
+    struct shared_hstring_header : hstring_header
+    {
+        atomic_ref_count count;
+        wchar_t buffer[1];
+    };
+
+    inline void release_hstring(hstring_header* handle) noexcept
+    {
+        WINRT_ASSERT((handle->flags & hstring_reference_flag) == 0);
+
+        if (0 == --static_cast<shared_hstring_header*>(handle)->count)
+        {
+            WINRT_HeapFree(WINRT_GetProcessHeap(), 0, handle);
+        }
     }
 
-    inline void* create_string(wchar_t const* value, uint32_t const length)
+    inline shared_hstring_header* precreate_hstring_on_heap(uint32_t length)
     {
-        void* result = nullptr;
-        check_hresult(WINRT_WindowsCreateString(value, length, &result));
-        return result;
+        WINRT_ASSERT(length != 0);
+
+        auto header = static_cast<shared_hstring_header*>(WINRT_HeapAlloc(WINRT_GetProcessHeap(), 0, sizeof(shared_hstring_header) + sizeof(wchar_t) * length));
+
+        if (!header)
+        {
+            throw std::bad_alloc();
+        }
+
+        header->flags = 0;
+        header->length = length;
+        header->ptr = header->buffer;
+        header->count = 1;
+        header->buffer[length] = 0;
+        return header;
+    }
+
+    inline hstring_header* create_hstring_on_heap(wchar_t const* value, uint32_t length)
+    {
+        if (!length)
+        {
+            return nullptr;
+        }
+
+        auto header = precreate_hstring_on_heap(length);
+        memcpy(header->buffer, value, sizeof(wchar_t) * length);
```

Comment:
I can't think of a reason to use `memcpy` instead of `std::copy` or `std::copy_n`.

Reply:
👍